File: org.eclipse.scout.rt.client/src/org/eclipse/scout/rt/client/ui/form/fields/groupbox/internal/GroupBoxBodyGrid.java
/*******************************************************************************
 * Copyright (c) 2010 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 ******************************************************************************/
package org.eclipse.scout.rt.client.ui.form.fields.groupbox.internal;

import java.util.ArrayList;

import org.eclipse.scout.commons.logger.IScoutLogger;
import org.eclipse.scout.commons.logger.ScoutLogManager;
import org.eclipse.scout.rt.client.ui.form.fields.GridData;
import org.eclipse.scout.rt.client.ui.form.fields.IFormField;
import org.eclipse.scout.rt.client.ui.form.fields.button.IButton;
import org.eclipse.scout.rt.client.ui.form.fields.groupbox.IGroupBox;
import org.eclipse.scout.rt.client.ui.form.fields.internal.GridDataBuilder;

/**
 * Grid (model) layout of items only visible items and non-process-buttons are
 * used
 */
public class GroupBoxBodyGrid {
  private static final IScoutLogger LOG = ScoutLogManager.getLogger(GroupBoxBodyGrid.class);

  private final IGroupBox m_groupBox;
  private IFormField[] m_fieldsExceptProcessButtons;
  private int m_gridColumns;
  private int m_gridRows;

  public GroupBoxBodyGrid(IGroupBox groupBox) {
    m_groupBox = groupBox;
  }

  public void validate() {
    // reset old state
    m_gridRows = 0;
    // STEP 0: column count
    m_gridColumns = -1;
    IGroupBox tmp = m_groupBox;
    while (m_gridColumns < 0 && tmp != null) {
      m_gridColumns = tmp.getGridColumnCountHint();
      tmp = tmp.getParentGroupBox();
    }
    if (m_gridColumns < 0) {
      m_gridColumns = 2;
    }
    int containingGridXYCount = 0;
    int notContainingGridXYCount = 0;
    // build
    ArrayList<IFormField> fieldsExceptProcessButtons = new ArrayList<IFormField>();
    IFormField[] fields = m_groupBox.getFields();
    for (int i = 0; i < fields.length; i++) {
      IFormField f = fields[i];
      if (f.isVisible()) {
        if ((f instanceof IButton) && (((IButton) f).isProcessButton())) {
          // ignore
        }
        else {
          fieldsExceptProcessButtons.add(f);
          GridData hints = f.getGridDataHints();
          if (hints.x >= 0 && hints.y >= 0) {
            containingGridXYCount++;
          }
          else {
            notContainingGridXYCount++;
          }
        }
      }
      else {
        GridData data = GridDataBuilder.createFromHints(f, 1);
        f.setGridDataInternal(data);
      }
    }
    m_fieldsExceptProcessButtons = fieldsExceptProcessButtons.toArray(new IFormField[0]);
    boolean isContainingXAndY = (containingGridXYCount > 0 && notContainingGridXYCount == 0);
    if (isContainingXAndY) {
      layoutAllStatic();
    }
    else {
      layoutAllDynamic();
    }
  }

  /**
   * Make layout based on grid-x, grid-y, grid-w and grid-h No auto-layout
   */
  private void layoutAllStatic() {
    int totalGridW = 1;
    int totalGridH = 0;
    for (int i = 0; i < m_fieldsExceptProcessButtons.length; i++) {
      IFormField f = m_fieldsExceptProcessButtons[i];
      GridData hints = GridDataBuilder.createFromHints(f, 1);
      totalGridW = Math.max(totalGridW, hints.x + hints.w);
      totalGridH = Math.max(totalGridH, hints.y + hints.h);
    }
    for (int i = 0; i < m_fieldsExceptProcessButtons.length; i++) {
      IFormField f = m_fieldsExceptProcessButtons[i];
      GridData hints = GridDataBuilder.createFromHints(f, totalGridW);
      f.setGridDataInternal(hints);
    }
    m_gridColumns = totalGridW;
    m_gridRows = totalGridH;
  }

  /**
   * Make auto-layout based on grid-w and grid-h only. ignores grid-x and grid-y
   */
  private void layoutAllDynamic() {
    // STEP 1: component tree
    GridCell main = null;
    main = layoutControls(main);
    // STEP 2: logical layout
    if (main != null) {
      // set coordinates
      main.calculateGridLayout(0, 0);
      // store height
      m_gridRows = main.getMaxY();
      m_gridColumns = main.getMaxCols();
    }
  }

  private GridCell/* next main */layoutControls(GridCell main) {
    GridCell part = null;
    for (int i = 0; i < m_fieldsExceptProcessButtons.length; i++) {
      IFormField f = m_fieldsExceptProcessButtons[i];
      GridCell c = new GridCell(f, m_gridColumns);
      if (c.data.w >= m_gridColumns) {
        c.data.w = m_gridColumns;
        // layout part of 1-column items and attach to main grid
        if (part != null) {
          if (m_gridColumns > 1) {// split
            int k = m_gridColumns;
            GridCell splitRoot = part;
            while (k >= 2) {
              int maxWeight = splitRoot.getMaxWeight();
              GridCell split = splitRoot.getSplitCell(0, maxWeight, k);
              if (split != null && split != splitRoot) {
                split.getUp().setDown(null);
                int splitOffMaxCols = Math.max(1, splitRoot.getMaxCols());
                //maximum columns in part before split (normally 1)
                splitRoot.setRight(split);
                splitRoot = split;
                // reduce by maximum column count of split-off part
                k = k - splitOffMaxCols;
              }
              else {
                break;
              }
            }// end for k
          }
          if (main == null) {
            main = part;
          }
          else {
            main.addBottomCell(part);
          }
          part = null;
        }// end part!=null
        if (main == null) {
          main = c;
        }
        else {
          main.addBottomCell(c);
        }
      }// end c.w>=m_gridColumns
      else {
        if (part == null) {
          part = c;
        }
        else {
          part.addBottomCell(c);
        }
      }
    }
    // add and split rest
    if (part != null) {
      if (m_gridColumns > 1) {
        int k = m_gridColumns;
        GridCell splitRoot = part;
        while (k >= 2) {
          int maxWeight = splitRoot.getMaxWeight();
          GridCell split = splitRoot.getSplitCell(0, maxWeight, k);
          if (split != null && split != splitRoot) {
            split.getUp().setDown(null);
            int splitOffMaxCols = Math.max(1, splitRoot.getMaxCols());
            // maximum columns in part before split (normally 1)
            splitRoot.setRight(split);
            splitRoot = split;
            // reduce by maximum column count of split-off part
            k = k - splitOffMaxCols;
          }
          else {
            break;
          }
        }// end for k
      }
      if (main == null) {
        main = part;
      }
      else {
        main.addBottomCell(part);
      }
      part = null;
    }
    return main;
  }

  public int getGridColumnCount() {
    return m_gridColumns;
  }

  public int getGridRowCount() {
    return m_gridRows;
  }

  @Override
  public String toString() {
    return getClass().getSimpleName() + "[" + getGridRowCount() + " " + getGridColumnCount() + "]";
  }
}

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/form/fields/groupbox/grid/GroupBoxBodyGrid.js
/*******************************************************************************
 * Copyright (c) 2014-2017 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 ******************************************************************************/
/**
 * @abstract
 */
scout.GroupBoxBodyGrid = function() {
  scout.GroupBoxBodyGrid.parent.call(this);
  this.gridRows = 0;
  this.gridColumns = 0;
};
scout.inherits(scout.GroupBoxBodyGrid, scout.LogicalGrid);

/**
 * @override
 */
scout.GroupBoxBodyGrid.prototype._validate = function(groupBox) {
  // reset old state
  this.gridRows = 0;
  // step 0: column count
  this.gridColumns = this._computeGridColumnCount(groupBox);
  var containingGridXYCount = 0;
  var notContainingGridXYCount = 0;
  // build
  var fieldsExceptProcessButtons = [];
  groupBox.fields.forEach(function(field) {
    if (field.visible) {
      if (!field.processButton) {
        fieldsExceptProcessButtons.push(field);
        var hints = field.gridDataHints;
        if (hints && hints.x >= 0 && hints.y >= 0) {
          containingGridXYCount++;
        } else {
          notContainingGridXYCount++;
        }
      }
    } else {
      var gd = scout.GridData.createFromHints(field, 1);
      field.gridData = gd;
    }
  }.bind(this));
  if (containingGridXYCount > 0 && notContainingGridXYCount === 0) {
    this.layoutAllStatic(fieldsExceptProcessButtons);
  } else {
    this.layoutAllDynamic(fieldsExceptProcessButtons);
  }
};

scout.GroupBoxBodyGrid.prototype.layoutAllStatic = function(fields) {
  var hints = [];
  fields.forEach(function(v) {
    hints.push(scout.GridData.createFromHints(v, 1));
  });
  var totalGridW = hints.reduce(function(x, y) {
    var y1 = y.x + y.w;
    return y1 > x ? y1 : x;
  }, 1);
  var totalGridH = hints.reduce(function(x, y) {
    var y1 = y.y + y.h;
    return y1 > x ? y1 : x;
  }, 0);
  fields.forEach(function(v) {
    v.gridData = scout.GridData.createFromHints(v, totalGridW);
  });
  this.gridRows = totalGridH;
  this.gridColumns = totalGridW;
};

scout.GroupBoxBodyGrid.prototype.layoutAllDynamic = function(fields) {
  // abstract, must be implemented by sub classes
};

scout.GroupBoxBodyGrid.prototype.getGridColumnCount = function() {
  return this.gridColumns;
};

scout.GroupBoxBodyGrid.prototype.getGridRowCount = function() {
  return this.gridRows;
};

scout.GroupBoxBodyGrid.prototype._computeGridColumnCount = function(groupBox) {
  var gridColumns = -1,
    tmp = groupBox;
  do {
    gridColumns = scout.nvl(tmp.gridColumnCount, gridColumns);
  } while (gridColumns < 0 && tmp.getParentGroupBox && (tmp = tmp.getParentGroupBox()));
  return gridColumns < 0 ? 2 : gridColumns;
};

/**
 * if grid w greater than group box column count. Grid w will be set to group box column count.
 */
scout.GroupBoxBodyGrid.getGridDataFromHints = function(field, groupBoxColumnCount) {
  var data = scout.GridData.createFromHints(field, groupBoxColumnCount);
  data.w = Math.min(groupBoxColumnCount, data.w);
  return data;
};
